agent/engine/containermetadata/write_metadata.go

```diff
@@ -0,0 +1,242 @@
+func getMetadataFilePath(task *api.Task, container *api.Container, dataDir string) string {
+	taskID := getTaskIDfromArn(task.Arn)
+	// Empty task ID indicates malformed Arn (Should not happen)
+	if taskID == "" {
+		return ""
+	}
+	return fmt.Sprintf("%s/metadata/%s/%s/", dataDir, taskID, container.Name)
+}
+
+// mdFileExist checks if metadata file exists or not
+func mdFileExist(task *api.Task, container *api.Container, dataDir string) bool {
+	mdFileDir := getMetadataFilePath(task, container, dataDir)
+	mdFilePath := fmt.Sprintf("%s/%s", mdFileDir, metadataFile)
+	if _, err := os.Stat(mdFilePath); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
+// writeToMetadata puts the metadata into JSON format and writes into
+// the metadata file
+func (md *Metadata) writeToMetadataFile(task *api.Task, container *api.Container, dataDir string) error {
+	data, err := json.MarshalIndent(md, "", "\t")
+	if err != nil {
+		return err
+	}
+	mdFileDir := getMetadataFilePath(task, container, dataDir)
+	mdFilePath := fmt.Sprintf("%s/%s", mdFileDir, metadataFile)
+
+	mdFile, err := os.OpenFile(mdFilePath, os.O_WRONLY, 0644)
+	defer mdFile.Close()
+	if err != nil {
+		return err
+	}
+	_, err = mdFile.Write(data)
+	return err
+}
+
+// getTaskMetadataDir acquires the directory with all of the metadata
+// files of a given task
+func getTaskMetadataDir(task *api.Task, dataDir string) string {
+	return fmt.Sprintf("%s/metadata/%s/", dataDir, getTaskIDfromArn(task.Arn))
+}
+
+// removeContents removes a directory and all its children. We use this
+// instead of os.RemoveAll to handle case where the directory does not exist
+func removeContents(dir string) error {
+	d, err := os.Open(dir)
+	if err != nil {
+		return err
+	}
+	defer d.Close()
+	names, err := d.Readdirnames(-1)
+	if err != nil {
+		return nil
+	}
+	for _, name := range names {
+		err = os.RemoveAll(filepath.Join(dir, name))
+		if err != nil {
+			return err
+		}
+	}
+	return os.Remove(dir)
+}
+
+// MetadataManager is an interface that allows us to abstract away the metadata
+// operations
+type MetadataManager interface {
+	CreateMetadata(*[]string, *api.Task, *api.Container) error
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client dockerDummyClient
+	cfg    *config.Config
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerDummyClient, cfg *config.Config) *metadataManager {
+	manager := &metadataManager{
+		client: client,
+		cfg:    cfg,
+	}
+	return manager
+}
+
+// CreateMetadata creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+func (manager *metadataManager) CreateMetadata(binds *[]string, task *api.Task, container *api.Container) error {
```

Comment:
Although Go passes slices as references, it appears that updating the slice, such as appending, does not change the address the slice points to so []string will still point to the old location. The solutions to this was either return a []string and pass by value, or pass by reference a *[]string to actually change the pointer as well.